ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_CppSamples/ZipHelper.cpp

```diff
@@ -0,0 +1,715 @@
+}
+
+void ZipHelper::zwClose(const QString& globalComment)
+{
+    if (m_zipFile)
+    {
+        zipClose(m_zipFile, qPrintable(globalComment));
+        m_zipFile = nullptr;
+    }
+}
+
+//--------------------------------------------------------------------
+
+/*!
+  \qmlmethod void ZipArchive::exists()
+  \brief Gets whether archive exists.
+  \internal
+*/
+
+bool ZipHelper::exists() const
+{
+    if (path().isEmpty())
+    {
+        return false;
+    }
+
+    return QFileInfo::exists(path());
+}
+
+//--------------------------------------------------------------------
+
+bool ZipHelper::zrOpen()
+{
+    if (zrIsOpen())
+    {
+        return true;
+    }
+
+    zwClose();
+
+    if (!exists())
+    {
+        return false;
+    }
+
+    auto ok = zrOpen(path());
+
+    if (!ok)
+    {
+#ifdef ZIP_LOGGING_ENABLED
+        qDebug() << "Unable to open:" << path();
+#endif
+    }
+
+    return ok;
+}
+
+void ZipHelper::setPath(const QString& path)
+{
+    auto resolvedPath = ZipHelper::resolvedPath(path);
+
+    if (resolvedPath == m_Path)
+    {
+        return;
+    }
+
+    m_Path = resolvedPath;
+
+    zwClose();
+    zrClose();
+
+    emit pathChanged();
+}
+
+//--------------------------------------------------------------------
+/*!
+  \qmlmethod bool ZipArchive::extractAll(string outputPath)
+  \brief Extracts all files from the archive.
+  \list
+    \li \a outputPath The path to the destination folder.
+  \endlist
+
+  Returns \c true if successful; otherwise \c false.
+*/
+
+bool ZipHelper::extractAll(const QString &outputPath)
+{
+    if (!zrOpen())
+    {
+#ifdef ZIP_LOGGING_ENABLED
+        qDebug() << "ZIP not open" << path();
+#endif
+        return false;
+    }
+
+    QDir outputDir(outputPath);
+
+#ifdef ZIP_LOGGING_ENABLED
+    qDebug() << outputPath << outputDir;
+#endif
+
+    return extractAll(outputDir);
+}
+
+// Methods moved from FileUtility to enable ZipHelper::setPath to work
+//--------------------------------------------------------------------
+
+#define kResourceScheme                         QStringLiteral("qrc")
+#define kResourcePathPrefix                     QStringLiteral(":/")
+
+#define kHttpScheme                             QStringLiteral("http")
+#define kHttpsScheme                            QStringLiteral("https")
+
+#define kPathHomePrefix                         QStringLiteral("~")
+#define kPathVarPrefix                          QStringLiteral("$")
+#define kPathVarHome                            QStringLiteral("HOME")
+#define kPathVarTemp                            QStringLiteral("TEMP")
+
+#define kGroupName                              "Esri/ArcGISRuntime/pathVariables"
+
+#define kPropertyPathVariablePrefix             kGroupName "."
+
+#define kSettingsGroupPathVariables             QStringLiteral(kGroupName)
+
+//------------------------------------------------------------------------------
+
+QString ZipHelper::resolvedPath(const QString& path, bool useFallbacks)
```

Comment:
Is stuff like this called anywhere? It doesn't look like zip functionality.

Reply:
Yeah it's called in `ZipHelper::setPath`.